src/transports/ws/sws.c

```diff
@@ -1298,8 +1379,19 @@ static void nn_sws_handler (struct nn_fsm *self, int src, int type,
                         sws->inmsg_current_chunk_buf = sws->inmsg_control;
                     }
                     else {
-                        sws->inmsg_chunks++;
                         sws->inmsg_total_size += sws->inmsg_current_chunk_len;
+                        /*  Protect non-control messages against the
+                            NN_RCVMAXSIZE threshold; control messages already
+                            have a small pre-allocated buffer, and therefore
+                            are not subject to this limit. */
+                        nn_pipebase_getopt (&sws->pipebase, NN_SOL_SOCKET,
+                            NN_RCVMAXSIZE, &opt, &opt_sz);
+                        if (opt >= 0 && sws->inmsg_total_size > (size_t) opt) {
+                            nn_sws_fail_conn (sws, NN_SWS_CLOSE_ERR_TOOBIG,
+                                "Message larger than application allows.");
```

Comment:
This message, if exposed to users, could be confusing.  "Message larger than NN_RCVMAXSIZE" might be clearer.  Or just "message size exceeds limit"?